File: KIJUNG/PRO_WEEK_06/프렌즈4블록.cpp
#include <string>
#include <vector>
#include <deque>
#include <algorithm>

using namespace std;

int checking[30][30];

void check(int row, int col, vector<deque<char>> board2){
    if(col+1>board2[row].size() || col+1>board2[row+1].size()) return;
    if(board2[row][col] == board2[row+1][col] &&
        board2[row][col] == board2[row][col+1] &&
        board2[row][col] == board2[row+1][col+1]){
            checking[row][col] = 1;
            checking[row+1][col] = 1;
            checking[row][col+1] = 1;
            checking[row+1][col+1] = 1;
        }
}

int solution(int m, int n, vector<string> board) {
    int answer = 0;
    vector<deque<char>> board2;

    for(int i=0;i<n;i++){
        deque<char> tmp;
        for(int j=m-1;j>=0;j--){
            tmp.push_back(board[j][i]);
        }
        board2.push_back(tmp);
    }

    while(1){
        fill(&checking[0][0], &checking[29][30], 0);
        
        for(int i=0;i<n-1;i++){
            for(int j=0;j<m-1;j++){
                check(i,j,board2);
            }
        }
        
        if(count(&checking[0][0], &checking[29][30], 1) == 0) break;
        answer += count(&checking[0][0], &checking[29][30], 1);
        
        for(int i=0;i<n;i++){
            int tmpSize = board2[i].size();
            for(int j=0;j<tmpSize;j++){
                if(checking[i][j] == 0) board2[i].push_back(board2[i].front());
                board2[i].pop_front();
            }
        }
    }

    return answer;
}

File: DONGUK/프로그래머스_WEEK 7/프렌즈4블록.py
def check(i,j, board):
    c = board[i][j]

    if c == '.':
        return False
    
    if  c == board[i][j+1] and c == board[i+1][j] and c == board[i+1][j+1]:
        return True
    else:
        return False    

def solution(m, n, board):
    answer = 0
    # 전처리
    for i in range(len(board)):
        board[i] = list(board[i])

    while True:
        checked = [[0] * n for _ in range(m)]

        # 같은 캐릭터 체크
        for i in range(m-1):
            for j in range(n-1):
                if check(i,j,board):
                    checked[i][j] = 1
                    checked[i+1][j] = 1
                    checked[i][j+1] = 1
                    checked[i+1][j+1] = 1

        cnt = 0
        for data in checked:
            cnt += data.count(1)  

        if cnt == 0:
            break

        answer += cnt

        for j in range(n):
            for i in range(m):
                if checked[i][j] == 1:
                    x, y = i, j
                    while x >= 0:
                        if x == 0:
                            board[x][y] = '.'
                            break
                        else:
                            board[x][y] = board[x-1][y]
                            x = x - 1
    return answer